src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/Models/InputData.cs

```diff
@@ -9,17 +9,60 @@ namespace RegressionGames.StateRecorder.Models
 {
     [Serializable]
     [SuppressMessage("ReSharper", "InconsistentNaming")]
-    public class InputData: IComponentDataProvider
+    public class InputData : IComponentDataProvider
     {
         // version of this schema, update this if fields change
         public int apiVersion = SdkApiVersion.VERSION_1;
+
         public int ApiVersion()
         {
             return apiVersion;
         }
+
         public List<KeyboardInputActionData> keyboard;
         public List<MouseInputActionData> mouse;
 
+        private List<object> allInputsSorted = null;
+
+        public List<object> AllInputsSortedByTime() {
+            if (allInputsSorted == null)
+            {
+                allInputsSorted = new();
+                allInputsSorted.AddRange(keyboard);
+
+                foreach (var mouseInputActionData in mouse)
+                {
+                    var allInputsSortedCount = allInputsSorted.Count;
+                    for (var i = 0; i < allInputsSortedCount; i++)
+                    {
+                        var input = allInputsSorted[i];
+                        if (input is KeyboardInputActionData keyboardData)
+                        {
+                            // if a keyboard entry time is after the new mouse time.. insert the mouse before this one
+                            if ((keyboardData.startTime.HasValue && keyboardData.startTime.Value > mouseInputActionData.startTime) ||
+                                (keyboardData.endTime.HasValue && keyboardData.endTime.Value > mouseInputActionData.startTime))
+                            {
+                                allInputsSorted.Insert(i, mouseInputActionData);
+                                break;
+                            }
+                        }
+                        else if (input is MouseInputActionData mouseData)
+                        {
+                            // if an existing mouse time is after the new mouse time.. insert the new mouse before this one
+                            if (mouseData.startTime > mouseInputActionData.startTime)
+                            {
+                                allInputsSorted.Insert(i, mouseInputActionData);
+                                break;
+                            }
+                        }
+                    }
+                    // didn't find something to put it before, stick it on the end
+                    allInputsSorted.Add(mouseInputActionData);
```

Comment:
I believe there's a bug here: This line will get called for every mouse input, even if a place for it is already found above.

Reply:
good catch, ty

Reply:
fixed